pkg/server/plugin/keymanager/hashicorpvault/hashicorp_vault.go

```diff
@@ -0,0 +1,490 @@
+			return 0, err
+		}
+		authMethod = CERT
+	}
+
+	if config.K8sAuth != nil {
+		if err := checkForAuthMethodConfigured(authMethod); err != nil {
+			return 0, err
+		}
+		authMethod = K8S
+	}
+
+	if authMethod != 0 {
+		return authMethod, nil
+	}
+
+	return 0, status.Error(codes.InvalidArgument, "one of the available authentication methods must be configured: 'Token, AppRole'")
+}
+
+func checkForAuthMethodConfigured(authMethod AuthMethod) error {
+	if authMethod != 0 {
+		return status.Error(codes.InvalidArgument, "only one authentication method can be configured")
+	}
+	return nil
+}
+
+func (p *Plugin) genClientParams(method AuthMethod, config *Config) (*ClientParams, error) {
+	cp := &ClientParams{
+		VaultAddr:         p.getEnvOrDefault(envVaultAddr, config.VaultAddr),
+		Namespace:         p.getEnvOrDefault(envVaultNamespace, config.Namespace),
+		TransitEnginePath: p.getEnvOrDefault(envVaultTransitEnginePath, config.TransitEnginePath),
+		CACertPath:        p.getEnvOrDefault(envVaultCACert, config.CACertPath),
+		TLSSKipVerify:     config.InsecureSkipVerify,
+	}
+
+	switch method {
+	case TOKEN:
+		cp.Token = p.getEnvOrDefault(envVaultToken, config.TokenAuth.Token)
+	case APPROLE:
+		cp.AppRoleAuthMountPoint = config.AppRoleAuth.AppRoleMountPoint
+		cp.AppRoleID = p.getEnvOrDefault(envVaultAppRoleID, config.AppRoleAuth.RoleID)
+		cp.AppRoleSecretID = p.getEnvOrDefault(envVaultAppRoleSecretID, config.AppRoleAuth.SecretID)
+	case CERT:
+		cp.CertAuthMountPoint = config.CertAuth.CertAuthMountPoint
+		cp.CertAuthRoleName = config.CertAuth.CertAuthRoleName
+		cp.ClientCertPath = p.getEnvOrDefault(envVaultClientCert, config.CertAuth.ClientCertPath)
+		cp.ClientKeyPath = p.getEnvOrDefault(envVaultClientKey, config.CertAuth.ClientKeyPath)
+	case K8S:
+		if config.K8sAuth.K8sAuthRoleName == "" {
+			return nil, status.Error(codes.InvalidArgument, "k8s_auth_role_name is required")
+		}
+		if config.K8sAuth.TokenPath == "" {
+			return nil, status.Error(codes.InvalidArgument, "token_path is required")
+		}
+		cp.K8sAuthMountPoint = config.K8sAuth.K8sAuthMountPoint
+		cp.K8sAuthRoleName = config.K8sAuth.K8sAuthRoleName
+		cp.K8sAuthTokenPath = config.K8sAuth.TokenPath
+	}
+
+	return cp, nil
+}
+
+func (p *Plugin) getEnvOrDefault(envKey, fallback string) string {
+	if value, ok := p.hooks.lookupEnv(envKey); ok {
+		return value
+	}
+	return fallback
+}
+
+func (p *Plugin) GenerateKey(ctx context.Context, req *keymanagerv1.GenerateKeyRequest) (*keymanagerv1.GenerateKeyResponse, error) {
+	if req.KeyId == "" {
+		return nil, status.Error(codes.InvalidArgument, "key id is required")
+	}
+	if req.KeyType == keymanagerv1.KeyType_UNSPECIFIED_KEY_TYPE {
+		return nil, status.Error(codes.InvalidArgument, "key type is required")
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	spireKeyID := req.KeyId
+	newKeyEntry, err := p.createKey(ctx, spireKeyID, req.KeyType)
+	if err != nil {
+		return nil, err
+	}
+
+	p.entries[spireKeyID] = *newKeyEntry
```

Comment:
It seems that all keys with the same SPIRE Key ID have the same path.
I think that the logic here should be that if we already have this SPIRE Key ID cached, a new key is created and the old one should be deleted.
I've added a comment about my concern about the current implementation that doesn't seem to be creating a new key, since the same path is always used for a given SPIRE Key ID.

Reply:
I think that we should first create the key, and then schedule the old one for deletion if exists. Otherwise, if the deletion happens first and it fails, the key is deleted and there will be no key that replaces it, which is a pretty bad situation.

We need to consider also that in an HA deployment, multiple SPIRE servers will be accessing the same Vault instance, so there should be a way to identify each server. There should be a `KeyIdentifierValue` config attribute as we have in other key managers for this purpose. With that attribute, the server ID can be persisted and retrieved. The plugin needs that to know which keys are managed by this specific instance of SPIRE Server.

Reply:
@amartinezfayo I've implemented it as you suggested. Now, a key is scheduled for deletion and each key has a unique id associated to it.

The `KeyIdentifierValue` config attribute is also there. Unfortunately, HashiCorp Vault doesn't support tags on keys. So we need a different way to associate keys with SPIRE servers. Perhaps a clever naming schema could solve this problem.

Reply:
>The KeyIdentifierValue config attribute is also there. Unfortunately, HashiCorp Vault doesn't support tags on keys.

This is unfortunate, because it makes difficult the identification and disposal of stale keys.
I think that we can embed the KeyIdentifierValue as part of the key name. I also think that it would be useful to be able to identify all the keys belonging to a trust domain, but the user could do that through the use of specific transit engine paths for each trust domain.
Prefixing the key names with "spire-key" is probably a good idea, to easily identify that they are SPIRE keys.